src/sayfudinova/task5/main.cpp

```diff
@@ -0,0 +1,195 @@
+#include "Light.h"
+#include "Plane.h"
+#include "Ray.h"
+#include "Sphere.h"
+
+#include <QCommandLineParser>
+#include <QCoreApplication>
+#include <QImage>
+
+constexpr float PI = 3.141592f;
+constexpr float FOV = PI / 3; // field of view angle
+auto WIDTH = 1024;
+auto HEIGHT = 720;
+constexpr auto REFLECTION_DEPTH = 4;
+constexpr QVector3D BACKGROUND_COLOR = QVector3D(0.9f, 0.9f, 0.9f);
+constexpr QVector3D LIGHT_COLOR{1.0f, 1.0f, 1.0f};
+
+inline QVector3D mix(const QVector3D &a, const QVector3D &b, const float k) {
+  return a * k + b * (1 - k);
+}
+
+QVector3D reflect(const QVector3D &incident, const QVector3D &normal) {
+  return incident - 2.f * normal * (QVector3D::dotProduct(incident, normal));
+}
+
+// Snell's law
+QVector3D refract(const QVector3D &incident, const QVector3D &normal, const float n_1, const float n_2 = 1.f) {
+  auto cos_1 = -std::max(-1.f, std::min(1.f, QVector3D::dotProduct(incident, normal)));
+  if (cos_1 < 0.f)
+    return refract(incident, -normal, n_2, n_1);
+  auto ratio = n_2 / n_1;
+  auto cos_2 = sqrtf(1 - ratio * ratio * (1 - cos_1 * cos_1));
+  return ratio * incident + (ratio * cos_1 - cos_2) * normal;
+}
+
+bool scene_intersect(const Ray &ray, const std::vector<Sphere> &spheres, const std::vector<Plane> &planes, QVector3D &hit, QVector3D &normal, Material &material) {
+  auto spheres_dist = std::numeric_limits<float>::max();
+  for (size_t i = 0; i < spheres.size(); i++) {
+    auto dist_i = std::numeric_limits<float>::max();
+    if (spheres[i].ray_intersect(ray, dist_i) && dist_i < spheres_dist) {
+      spheres_dist = dist_i;
+      hit = ray.origin + ray.direction * dist_i;
+      normal = (hit - spheres[i].center).normalized();
+      material = spheres[i].material;
+    }
+  }
+
+  auto checkerboard_dist = std::numeric_limits<float>::max();
+  for (size_t i = 0; i < planes.size(); i++) {
+    auto dist_i = std::numeric_limits<float>::max();
+    if (planes[i].ray_intersect(ray, dist_i) && dist_i < spheres_dist) {
+      checkerboard_dist = dist_i;
+      hit = ray.origin + ray.direction * dist_i;
+      normal = planes[i].normal;
+      material.diffuseColor = (int(.5 * hit.x() + 10.) + int(.5 * hit.z())) & 1
+                                  ? QVector3D(.3f, .0f, .5f)
+                                  : QVector3D(.0f, .0f, .0f);
+    }
+  }
+  return std::min(spheres_dist, checkerboard_dist) < 1000.f;
+}
+
+QVector3D cast_ray(const Ray &ray, const std::vector<Sphere> &spheres,
+                   const std::vector<Plane> &planes,
+                   const std::vector<Light> &lights, size_t depth = 0) {
+  QVector3D point;
+  QVector3D normal;
+  Material material;
+
+  if (depth > REFLECTION_DEPTH || !scene_intersect(ray, spheres, planes, point, normal, material)) {
+    return BACKGROUND_COLOR;
+  }
+
+  QVector3D reflect_direction = reflect(ray.direction, normal).normalized();
```

Comment:
```suggestion
  const auto reflect_direction = reflect(ray.direction, normal).normalized();
```
и ниже где можно `const auto` или `auto`